<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
  <head>
	<title>MOA - Makefiles in bioinformatics</title>
	<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
	<link href="moa.css" rel="stylesheet" type="text/css">
	<link rel="icon" type="image/ico" href="favicon.ico">
    
  </head>
  <body>
	<center>
	  <div id="wrapper">
		<div style="width: 815px;">
		  <div style="clear: left;">
			<div style="float: left; vertical-align:text-bottom">
			  <img id="menulogo" align="middle" src="images/moa_logo_title.png">
			</div>
			<div style="float: right;">
			  <a href="http://www.plantandfood.co.nz/">
				<img src="images/pfr_logo.jpg" alt="" height="87" width="351" border="0">
			  </a>
			</div>
		  </div>
		  <div style="clear: left;">
			<div style="float: left; width: 715px; height: 31px; background-color: rgb(81, 0, 45);">
			  &nbsp;&nbsp;
			</div>
			<div style="float: left;">
			  <img src="images/scoop.jpg" alt="" height="31" width="100" border="0">
			</div>
		  </div>
		</div>
		<div style="clear: left;">
		  <div id="left">
			<div style="width: 160px;">
			  <div style="text-align: right;">
                
                <a href="introduction.html">Introduction</a><br>
                
                <a href="installation.html">Installation</a><br>
                
                <a href="commands.html">Commands</a><br>
                
                <a href="templates.html">Templates</a><br>
                
                <a href="extending.html">Extending</a><br>
                
                <a href="reference.html">Reference</a><br>
                
			  </div>
			</div>
		  </div>
		  <div id="center_content">
			<div id="homepage_container">
              
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <title></title>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="date" content="" />
</head>
<body>
<p
><strong
  >NOTE: both the software and the manual are under development. Expect things to change.</strong
  ></p
><div id="introduction"
><h1
  >Introduction</h1
  ><p
  >A bioinformatics project is often consists of many separate, interdependent, steps. Once a project gets bigger it becomes vital to apply some organization and automation.</p
  ><p
  >There are many different ways to automate a bioinformatics project . Many bioinformaticians, however, continue using the command-line or tailor-made scripts to organize and automate their work. This has obvious advantages, most importantly flexibility. Potential downsides to scripting are that a project easily becomes disorganized and untrackable unless explicit measures are taken.</p
  ><p
  >Moa aims to assist in organizing, automating and maintaining a command line bioinformatics project without losing flexibility. The best way to understand how Moa aims to achieve this is by an example:</p
  ><pre
  ><code
    >moa new blast -t &quot;run a demo&quot;
moa set blast_db=/data/blast/db/nt
moa set blast_input_dir=../sequences/
moa set moa_postprocess='grep polymerase gff/*gff &gt; pol.gff'
moa run
</code
    ></pre
  ><p
  >In the first line, a new BLAST  job (titled &quot;run demo&quot;) is created in the current directory. What really happens is that Moa creates a script (Makefile) that imports a BLAST specific template. The next two lines set a few parameters for BLAST. In the last line Moa is executed and 'blasts' the input sequences (in <code
    >../sequences</code
    >) against the database in <code
    >/data/blast/db/nt</code
    >. BLAST output files (XML) are generated and converted to GFF (GFF conversion is an extra, not part of the BLAST suite). The one to last statement is probably most remarkable; it is a single (shell) commmand that will be executed after BLAST is executed (there is a corresponding <code
    >moa_preprocess</code
    >). This shell comamand filters all BLAST hits that have the word &quot;polymerase&quot; in their description into a separater GFF file.</p
  ><p
  >Moa employs <a href="http://www.gnu.org/software/make"
    >GNU make</a
    > to describe its building blocks. GNU Make is originally developed for software compilation. Compilation usually involves the execution of many interdependent compilation and linker steps. GNU make is able to compile large software projects with tens of thousands of source files based on a Makefile that describes how a target file is to be created. GNU Make is flexible enough to be used with practically any programming language. Moreover, GNU Make can be used to automate any series of commands as long as they can be executed from the command line. It is therefore not only possible, but an excellent idea (not mine), to use Gnu Make in bioinformatics projects (see <a href="http://biowiki.org/MakefileManifesto"
    >biowiki</a
    >, <a href="http://archive.nodalpoint.org/2007/03/18/a_pipeline_is_a_makefile"
    >nodalpoint</a
    > or <a href="http://skam.sourceforge.net/"
    >biomake</a
    >)</p
  ><p
  >Moa aims to do the following things:</p
  ><ul
  ><li
    ><em
      >Organize a project</em
      >: Each Moa job must be located in its own directory. It is possible to automatically execute a directory tree of Moa jobs. Proper use of these features will result in a logical project structure.</li
    ><li
    ><em
      >Reuse building blocks</em
      >: Moa templates are GNU Makefiles that follow a set of conventions. It is easy to implement new building blocks. (see chapter X).</li
    ><li
    ><em
      >Document</em
      >: It is possible to add meta-data such as a title and description to each Moa job, making it easy to</li
    ><li
    ><em
      >Provide a uniform interface</em
      >: Moa allows you to operate your project almost exclusively using a single command (conveniently called <code
      >moa</code
      >).</li
    ></ul
  ><div id="example-session"
  ><h2
    >Example session</h2
    ><p
    >The best way to understand how to use Moa is a sample session.</p
    ><p
    >We'll start by creating directories to hold the data and analysis structure:</p
    ><pre
    ><code
      >mkdir introduction
cd introduction
</code
      ></pre
    ><p
    >We've created a directory <code
      >introduction</code
      > for the tutorial. Within this directory we'll organize the components of our analysis. We want to initialize this directory so that it becomes a part of this Moa pipeline. This is useful later, if we want to run all analysis at once. To do this, run:</p
    ><pre
    ><code
      >moa new project -t 'Introduction'
</code
      ></pre
    ><p
    >The <code
      >moa new</code
      > command is used to create new moa jobs. In this case we create a job with the template &quot;project&quot;. In itself this template does not do anything but serves to group new projects. The <code
      >-t</code
      > parameters assigns a title to this Moa job. We will now create a new directory to hold the first step of the pipeline:</p
    ><pre
    ><code
      >mkdir 10.download    
cd 10.download
moa new
</code
      ></pre
    ><p
    >Moa does not dictate a directory structure for your analysis pipeline, but to make full usage of Moa it is advisable to create a logical organization. Two important features of Moa that relate to this are:</p
    ><ul
    ><li
      >Each moa job is contained in one directory. Output files of a job are typically stored in that directory. It is not possible to have more than one Moa jobs in a directory.</li
      ><li
      >Moa is able to automatically execute all jobs in a directory and the underlying sub-directories (using <code
	>moa all</code
	>).</li
      ></ul
    ><p
    >If properly used, these two features force a logical, modular, project structure. To assign an order to the steps inside a directory it is possible to prepend a number to the directory name (i.e. &quot;10.&quot;). Note that Moa sorts directories alphabetically and not numerically</p
    ><p
    >We will now created a new folder to hold a genome sequence we are about to download and set up the Moa job to actually do the download.</p
    ><pre
    ><code
      >mkdir 10.genome
cd 10.genome
moa new -t 'download a potato BAC' ncbi
</code
      ></pre
    ><p
    >Here we create a Moa job to download a sequence from <a href="http://www.ncbi.nlm.nih.gov"
      >NCBI</a
      > by using the &quot;ncbi&quot; template. Once a Moa job is instantiated you can run <code
      >moa help</code
      > to get some information on how to use this template</p
    ><p
    ><img src="images/screenshot_moa_help.png" alt="moa help"
       /> Note that if you want help on how to use the moa itself, you should use <code
      >moa --help</code
      ></p
    ><p
    >Before you can get the data from NCBI, you will have to tell Moa what you want to download. This is easy if you know the Genbank accession numbers. In this case we'll download the nucleotide sequence (from the database &quot;nuccore&quot;) with the accession id AC237669.1</p
    ><pre
    ><code
      >moa set ncbi_db=nuccore 
moa set ncbi_query=AC237669.1
</code
      ></pre
    ><p
    >You can check if the parameters are set correctly by running <code
      >moa show</code
      >. This should come back with the following text: title download a potato BAC ncbi_query AC237669.1 ncbi_db nuccore ncbi_sequence_name</p
    ><p
    >If everything seems fine, you can run this job:</p
    ><pre
    ><code
      >moa
</code
      ></pre
    ><p
    >Or, you could also have used <code
      >moa run</code
      >. It is possible that you get an error message notifying that &quot;wget&quot; or &quot;xml_grep&quot; cannot be found. Most, properly written, Moa templates do prerequisite checking if necessary. If either of these tools is missing, you will need to install them first (possibly by running <code
      >sudo apt get install wget</code
      > or <code
      >sudo apt-get install xml-twig-tools</code
      >, or ask your system administrator)</p
    ><p
    >If Moa runs, quite a lot of output is generated. If things go wrong, there is probably a clue to why it did not work in this output. If the Moa job is successful, the last line should be &quot;Moa finished - Succes!&quot;. If you do an <code
      >ls</code
      > you now see a <code
      >fasta</code
      > directory with one fasta file. This fasta file contains the downloaded genome.</p
    ><p
    >Now we can start doing things with the downloaded sequence. To see what other templates are available, try <code
      >moa list</code
      >.</p
    ></div
  ></div
>
</body>
</html>



			</div>
          </div>          
          <p>&nbsp;</p>
          <div class="footer_text" id="footer">
            <small>This page is generated on Thu Aug 26 12:22:26 2010</small><br>
            &copy 2009 PLANT &amp; FOOD RESEARCH | 
            <a href="http://www.plantandfood.co.nz/disclaimer/">disclaimer</a>
          </div>
        </div>
      </div>
    </center>
  </body>
</html>
