<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
	<title>MOA - Makefiles in bioinformatics</title>
	<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
	<link href="moa.css" rel="stylesheet" type="text/css">
  </head>
  <body>
	<center>
	  <div id="wrapper">
		<div style="width: 815px;">
		  <div style="clear: left;">
			<div style="float: left; vertical-align:text-bottom">
			  <img id="menulogo" align="middle" src="moa_logo_title.png">
			</div>
			<div style="float: right;">
			  <a href="http://www.plantandfood.co.nz/">
				<img src="pfr_logo.jpg" alt="" height="87" width="351" border="0">
			</a>
			</div>
		  </div>
		  <div style="clear: left;">
			<div style="float: left; width: 715px; height: 31px; background-color: rgb(81, 0, 45);">
			  &nbsp;&nbsp;
			</div>
			<div style="float: left;">
			  <img src="scoop.jpg" alt="" height="31" width="100" border="0">
			</div>
		  </div>
		</div>
		<div style="clear: left;">
		  <div id="left">
			<div style="width: 160px;">
			  <div style="text-align: right;">
                 <a href="introduction.html">Introduction</a><br>
                 <a href="installation.html">Installation</a><br>
                 <a href="using.html">Usage</a><br>
                 <a href="gbrowse.html">Gbrowse</a><br>
                 <a href="extending.html">Extend Moa</a><br>
                 <a href="reference.html">Reference</a><br>
			  </div>
			</div>
		  </div>
		  <div id="center_content">
			<div id="homepage_container">
			
<h1 id="chapterTitle">Introduction</h1>
<p
><strong
  >NOTE:</strong
  > both the software and the manual are under heavy development. Expect things to be incomplete or change.</p
><p
>Moa is a set of tools build around <a href="http://www.gnu.org/software/make"
  >GNU make</a
  >  that facilitates the use of GNU make in bioinformatics data analysis.</p
><p
>GNU Make is developed to aid in compilation of software. Software compilation usually involves the execution of many preprocessing, compilation and linker steps, potentially with differing parameters and interdepending of each other.</p
><p
>Gnu make goes about a compilation process by describing what target files are to be created, on what source files or libraries the build is depending and how the target needs to be build. If a dependency needs to be compiled before it can be used for the final target file, its building process can be described as well (and so no). In big projects, this might result in thousands of files that are processed in the correct order. Moreover, Gnu Make is intelligent enough to repeat only the affected part of the build process if one or a few source files are changed.</p
><p
>The description that Gnu Make uses to execute software compilation is called a Makefile. The syntax of a Makefile is not only flexible enough to allow Gnu Make to be used with a wide variety of programming languages, but it can be used to automate any series of commands (as long as they can be executed from the command line). It is therefore a good idea to use Gnu Make in bioinformatics projects (see: )</p
><p
>A bioinformatics analysis often takes the form of a set of interdepending analyses, i.e..</p
><p
>...</p
><p
>Moa wraps a set of common bioinformatics tools as Makefiles. Using Moa gives you a:</p
><ul
><li
  ><p
    >A uniform interface; all Moa makefiles use a central library that provides a uniform, command line, interface to configuring and executing jobs.</p
    ></li
  ><li
  ><p
    >An easy way to track and repeat a set of analyses. Using Moa makes the creation of organized analysis structures easy.</p
    ></li
  ><li
  ><p
    >Interaction; templates are designed to interact with each other and make it easy to build pipelines using the Moa makefiles as building blocks.</p
    ></li
  ><li
  ><p
    >Parallel execution; Gnu make facillitates (limited) parallel execution of jobs. There is nothing however, that prevents integrations with a third party cluster solution such as Hadoop or SGE.</p
    ></li
  ></ul
><p
>Apart from a set of template Makefiles, the Moa contains several other</p
><ul
><li
  ><p
    >moaBase; a central library describing a number of central routines used by all Makefiles</p
    ></li
  ><li
  ><p
    >The &quot;moa&quot; helper script; a frontend to using Moa.</p
    ></li
  ><li
  ><p
    >Additional helper scripts; several of the template files require helper scripts that are part of the moa package.</p
    ></li
  ></ul
><div id="example-session"
><h2
  >Example session</h2
  ><p
  >To really understand how to use Moa, and how easy it is to do so, a sample session will be helpful.</p
  ><p
  >We'll start by creating directories to hold the data and analysis structure:</p
  ><pre
  ><code
    >mkdir introduction
cd introduction
mkdir 10.download
cd 10.download
</code
    ></pre
  ><p
  >We've created a directory called <code
    >sample</code
    > to store the complete analysis structure. Within this directory we'll start to organize the components of our sample analysis. Moa doesn't enforce a sequential organizaiton of an analysis pipeline, but expects the user to do this by using a logical directory structure. Hence, the directory describing the first step - downloading data from NCBI - is prefixed with a <code
    >10.</code
    >. Other, later steps will use higher numbers.</p
  ><pre
  ><code
    >mkdir 10.genome
</code
    ></pre
  ><p
  >In the <code
    >10.download</code
    > directory we repeat this; Create numbered directories to enforce organization. We'll now go to the <code
    >10.genome</code
    > directory and actually use Moa to download something from NCBI.</p
  ><pre
  ><code
    >cd 10.genome
moa new getFromNcbi
</code
    ></pre
  ><p
  >The utility script <code
    >moa</code
    > is used in occasions where direct interaction with GNU Make is not possible (yet). In this case the &quot;moa&quot; script creates our first Moa makefile. Whenever there is a Moa makefile you can type <code
    >make help</code
    >. Apart from some introduction, you'll get see some targets and parameters.</p
  ><p
  >Targets are things that a Moa makefile can do. You can use GNU Make to execute one of these target by running <code
    >Make TARGET</code
    >. Each Moa makefile usually has one main task (in this case, download data from NCBI). That main task is alway executed by running Make without a target defined (i.e. <code
    >make</code
    >).</p
  ><p
  >If you try to run <code
    >make</code
    > now, you'll get an error!. This is because you haven't told Moa yet what you want to download. You can do that by setting a few parameters. <code
    >make help</code
    > gives you an overview of all parameters that you can set. Moa distinguishes between required and optional parameters. Optional parameters can usually be guessed or are set to reasonable default values. In the case of the &quot;getFromNcbi&quot; Makefile, there are two parameters that really need to be set: <code
    >ncbi_db</code
    > and <code
    >ncbi_query</code
    >. In this case, it is possible to discover what these are by doing a query on the NCBI website and browse until you have the page with data you're looking for. In this case, we will are going to download a <strong
    ><em
      >Lactobacillus</em
      ></strong
    > genome from <a href="http://www.ncbi.nlm.nih.gov/nuccore/NC_012214"
    >NCBI</a
    >. Upon inspection of the URL (<code
    >http://www.ncbi.nlm.nih.gov/nuccore/NC_012214</code
    >) it is straightofrward identify the two parameters that need to be set:</p
  ><pre
  ><code
    >make set ncbi_db=nuccore
make set ncbi_query=NC_004567
</code
    ></pre
  ><p
  >Executing these two commands do not appear to do anything. You can, however, check what parameters are set by running:</p
  ><pre
  ><code
    >make show
</code
    ></pre
  ><p
  >You should now see:</p
  ><pre
  ><code
    >ncbi_db nuccore
ncbi_query  NC_004567
gfn_sequence_name   
jid moa_getFromNcbi_10.genome_??????????
project 
</code
    ></pre
  ><p
  >There are a few more parameters here, you can ignore all of these now, execpt for &quot;jid&quot;. A &quot;jid&quot;, or &quot;job id&quot; is a unique name that will be used to track information on this job. It is important, particularly if you're creating big projects, to set this to an understandable, descriptive, but short value (without spaces!!), so we'll run an additional <code
    >make set</code
    >:</p
  ><pre
  ><code
    >make set jid=lactobacillus.genome
</code
    ></pre
  ><p
  >This job is set up and can be executed now by running:</p
  ><pre
  ><code
    >make
</code
    ></pre
  ><p
  >This generates quite a lot of output. If things go wrong, there is probably a clue to why in the output. If the Moa job is successful, the last line should be &quot;Moa finished - Succes!&quot;. If you do an <code
    >ls</code
    > you now see a <code
    >fasta</code
    > directory with one fasta file. This fasta file contains the downloaded genome.</p
  ><p
  >Good, nice, but not very exciting. We'll now start doing something with this data. For example, we could map the results from a 454 run against this genome. Lets download a dataset from <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?"
    >NCBI's short read archive</a
    >:</p
  ><pre
  ><code
    >cd ..
mkdir 20.454reads
cd 20.454reads
mkdir 10.download
cd 10.download
moa new lftp
</code
    ></pre
  ><p
  >Here we use a different template: lftp. We need to set a single parameters (i.e. the location of the file to download).</p
  ><pre
  ><code
    >make set lftp_url=ftp://ftp.ncbi.nlm.nih.gov/sra/static/SRX002/SRX002375/SRR011133.fastq.gz
make set jid=lactobacillus.454.reads
</code
    ></pre
  ><p
  >again we can run <code
    >make</code
    > to download the data. After downloading we see that an archive is downloaded. We have now two options - unpack it manually or automate unpacking. We could do it manually. If the project is a quick, one-off, this is probably not so bad. However, you won't be able to completely rerun the analysis without having to think what steps you did manually. Therefore, we'll now set something up to also automate this.</p
  ><pre
  ><code
    >cd ..
mkdir 20.unpack
cd 20.unpack
moa new gather
</code
    ></pre
  ><p
  >The Moa &quot;gather&quot; template is probably not named very well. It's a very flexible template that allows you to do many different things on a set of input files (read the manual page). We'll set it up now to automate the unpacking.</p
  ><pre
  ><code
    >make set g_input_dir=../10.download
make set g_input_pattern=*.gz
</code
    ></pre
  ><p
  >These two lines define what files are used as input for the &quot;gather&quot; step. The next parameter is a sed command that strips of the .gz extension for the unpacked output file.</p
  ><pre
  ><code
    >make set g_name_sed='s/\.gz$//'
</code
    ></pre
  ><p
  >(Note that you must use ' quotes to prevent bash from expanding the $ sign as a variable!)</p
  ><pre
  ><code
    >make set g_process=`
</code
    ></pre
  ></div
>
</div>
</div>
  
<p>&nbsp;</p><div class="footer_text" id="footer">&copy 2009 PLANT &amp; FOOD RESEARCH | <a href="http://www.plantandfood.co.nz/disclaimer/">disclaimer</a></div>
</div>
</div></center>


</body>
</html>
