<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
	<title>MOA - Makefiles in bioinformatics</title>
	<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
	<link href="moa.css" rel="stylesheet" type="text/css">
	<link rel="icon" type="image/ico" href="favicon.ico">

  </head>
  <body>
	<center>
	  <div id="wrapper">
		<div style="width: 815px;">
		  <div style="clear: left;">
			<div style="float: left; vertical-align:text-bottom">
			  <img id="menulogo" align="middle" src="images/moa_logo_title.png">
			</div>
			<div style="float: right;">
			  <a href="http://www.plantandfood.co.nz/">
				<img src="images/pfr_logo.jpg" alt="" height="87" width="351" border="0">
			</a>
			</div>
		  </div>
		  <div style="clear: left;">
			<div style="float: left; width: 715px; height: 31px; background-color: rgb(81, 0, 45);">
			  &nbsp;&nbsp;
			</div>
			<div style="float: left;">
			  <img src="images/scoop.jpg" alt="" height="31" width="100" border="0">
			</div>
		  </div>
		</div>
		<div style="clear: left;">
		  <div id="left">
			<div style="width: 160px;">
			  <div style="text-align: right;">
                 <a href="introduction.html">Introduction</a><br>
                 <a href="installation.html">Installation</a><br>
                 <a href="extending.html">Extend Moa</a><br>
                 <a href="reference.html">Reference</a><br>
                 <hr>
                 <a href="moa_manual.pdf">Manual in PDF</a>
                 <hr>
                 <a href="api/index.html">Python API</a><br>
			  </div>
			</div>
		  </div>
		  <div id="center_content">
			<div id="homepage_container">
			
<h1 id="chapterTitle">getorf</h1>
<div id="targets"
><h1
  >Targets</h1
  ><dl
  ><dt
    >(empty)</dt
    ><dd
    >Execute the default target (getorf)</dd
    ><dt
    >getorf</dt
    ><dd
    >(Warning - a lazy developer refused to write a single line specifying what this template is supposed to do!!)</dd
    ><dt
    >clean</dt
    ><dd
    >removes all results from this job</dd
    ><dt
    >all</dt
    ><dd
    >executes the default target and into subdirectories to execute any other moa makefile it encounters</dd
    ></dl
  ></div
><div id="parameters"
><h1
  >Parameters</h1
  ><div id="required-parameters"
  ><h2
    >Required parameters</h2
    ><dl
    ><dt
      >title</dt
      ><dd
      >A job name - Describe what you are doing (string)</dd
      ><dt
      >getorf_input_dir</dt
      ><dd
      >Input files for getorf (directory)</dd
      ></dl
    ></div
  ><div id="optional-parameters"
  ><h2
    >Optional parameters</h2
    ><dl
    ><dt
      >moa_precommand</dt
      ><dd
      >A single command to be executed before the main operation starts. For more complicated processing, please override the moa_preprocess target in the local Makefile. (string)</dd
      ><dt
      >moa_postcommand</dt
      ><dd
      >A single shell command to be executed after the Moa is finished. For more complex processing please override the moa_postprocess target in the local Makefile. (string)</dd
      ><dt
      >getorf_input_extension</dt
      ><dd
      >file extension for the files in getorf_input_dir (string, default:fasta)</dd
      ><dt
      >getorf_input_glob</dt
      ><dd
      >glob to select a subset of files from getorf_input_dir (string, default:*)</dd
      ><dt
      >getorf_input_sort</dt
      ><dd
      >Sort order. Choose from: u - unsorted, s - size, sr - size reverse, t - time, tr - time reverse ()</dd
      ><dt
      >getorf_input_limit</dt
      ><dd
      >Number of files to use, if not defined: all files (integer)</dd
      ><dt
      >getorf_gff_source</dt
      ><dd
      >source field to use in the gff. (string, default:getorf)</dd
      ><dt
      >getorf_minsize</dt
      ><dd
      >minimal nucleotide size of the predicted ORF. (integer, default:30)</dd
      ><dt
      >getorf_maxsize</dt
      ><dd
      >maximal nucleotide size of the predicted ORF. (integer, default:1000000)</dd
      ><dt
      >getorf_circular</dt
      ><dd
      >Is the sequence linear? (Y|<strong
	>N</strong
	>)</dd
      ><dt
      >getorf_table</dt
      ><dd
      >Genetic code to use: 0 Standard; 1 Standard with alternative initiation codons; 2 Vertebrate Mitochondrial; 3 Yeast Mitochondrial; 4 Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma; 5 Invertebrate Mitochondrial; 6 Ciliate Macronuclear and Dasycladacean; 9 Echinoderm Mitochondrial; 10 Euplotid Nuclear; 11 Bacterial; 12 Alternative Yeast Nuclear; 13 Ascidian Mitochondrial; 14 Flatworm Mitochondrial; 15 Blepharisma Macronuclear; 16 Chlorophycean Mitochondrial; 21 Trematode Mitochondrial; 22 Scenedesmus obliquus; 23 Thraustochytrium Mitochondrial. (0|1|2|3|4|5|6|7|8|9|10|<strong
	>11</strong
	>|12|13|14|15|16|21|22|23)</dd
      ><dt
      >getorf_find</dt
      ><dd
      >What to output? 0: Translation between stop codons, 1: Translation between start &amp; stop codon, 2: Nucleotide sequence between stop codons; 3: Nucleotide sequence between start and stop codons. Default: 3 (0|1|2|<strong
	>3</strong
	>)</dd
      ></dl
    ></div
  ></div
>
</div>
</div>
  
<p>&nbsp;</p><div class="footer_text" id="footer">&copy 2009 PLANT &amp; FOOD RESEARCH | <a href="http://www.plantandfood.co.nz/disclaimer/">disclaimer</a></div>
</div>
</div></center>


</body>
</html>
