<html>
  <head>
	<title>Moa</title>
  </head>
  <link rel="stylesheet" href="moa.css">
  <body>
	<div id="menu">
	  <center>
		<a href="index.html">
		  <img id="menulogo" align="middle" src="moa_logo_smaller.png">
		</a>
	  </center>
	  <ul>
		<li>
		  <a href="introduction.html">Introduction</a>
		</li>
		<li>
		  <a href="installation.html">Installation</a>
		</li>
		<li>
		  <a href="using.html">Usage</a>
		</li>
		<li>
		  <a href="gbrowse.html">Gbrowse</a>
		</li>
		<li>
		  <a name="chapter_couchdb" href="couchdb.html">Couchdb</a>
		</li>
		<li>
		  <a href="extending.html">Extend Moa</a>
		</li>
		<li>
		  <a href="reference.html">Reference</a>
		</li>
	  </ul>
	  <small>
		<a href="manual.pdf">
		  download this manual in pdf format
		</a>
	  </small> 
   </div>
<div id="content">
<h1 id="chapterTitle">upload2gbrowse.mk</h1>
<p
>A library that aids in uploading FASTA and GFF to a Generic Genome Browser database. This template is only to be used embedded in another template. This library expects that the following variables are preset; gup_fasta_dir, gup_gff_dir, gffsource gup_upload_fasta, gup_upload_gff</p
><div id="targets"
><h1
  >Targets</h1
  ><dl
  ><dt
    >upload2gbrowse</dt
    ><dd
    ><p
      >Upload to gbrowse</p
      ></dd
    ><dt
    >initGbrowse</dt
    ><dd
    >Clean &amp; initalize a gbrowse database. <strong
      >Warning: all data will be lost!</strong
      ></dd
    ><dt
    >gupLock</dt
    ><dd
    >Prevent this job from uploading anything to the Generic Genome Browser database</dd
    ><dt
    >gupUnlock</dt
    ><dd
    ><p
      >Allow this job to upload to the Generic Genome Browser database</p
      ></dd
    ></dl
  ></div
><div id="parameters"
><h1
  >Parameters</h1
  ><div id="required-parameters"
  ><h2
    >Required parameters</h2
    ><dl
    ><dt
      >gup_gffsource</dt
      ><dd
      >the gff source field, used in batch operations</dd
      ><dt
      >gup_upload_fasta</dt
      ><dd
      >upload fasta to gbrowse (T/F)</dd
      ><dt
      >gup_upload_gff</dt
      ><dd
      >upload gff to gbrowse (T/F)</dd
      ><dt
      >gup_user</dt
      ><dd
      >gbrowse db user. If not defined, this defaults to 'moa'.</dd
      ><dt
      >gup_db</dt
      ><dd
      >gbrowse database. If not defined, this defaults to 'moa'.</dd
      ><dt
      >gup_gffsource</dt
      ><dd
      >the gff source field, used in batch operations</dd
      ></dl
    ></div
  ><div id="optional-parameters"
  ><h2
    >Optional parameters</h2
    ><dl
    ><dt
      >gup_fasta_dir</dt
      ><dd
      >input directory with fasta files to upload to gbrowse</dd
      ><dt
      >gup_gff_dir</dt
      ><dd
      >input directory with gff files to upload to gbrowse</dd
      ><dt
      >gup_gff_extension</dt
      ><dd
      >extension of the GFF files to upload (.gff)</dd
      ><dt
      >gup_fasta_extension</dt
      ><dd
      >extension of the FASTA files to upload (.fasta)</dd
      ><dt
      >gup_upload_fasta</dt
      ><dd
      >upload fasta to gbrowse (T/F)</dd
      ><dt
      >gup_upload_gff</dt
      ><dd
      >upload gff to gbrowse (T/F)</dd
      ><dt
      >gup_force_upload</dt
      ><dd
      >upload to gbrowse, ignore gup_lock and upload all, not only files newer that upload_gff or upload_fasta</dd
      ><dt
      >marks_extensions</dt
      ><dd
      >Add some extensions to the Gbrowse database to be initalized, for use by Mark.</dd
      ><dt
      >jid</dt
      ><dd
      >Unique identifier for this analysis job. Autogenerated unless defined.</dd
      ><dt
      >project</dt
      ><dd
      >undefined</dd
      ></dl
    ></div
  ></div
>
</div></body></html>
