\documentclass[a4paper,11pt,oneside]{article}

\renewcommand{\familydefault}{\sfdefault}
\usepackage[super]{natbib}

\begin{document}
\title{Annotation of the potato genome using GNU-make}
\author{Mark Fiers, Susan Thomson, Jeanne Jacobs}

Introduction

An increase in genome sequencing capacity requires a matching increase in
the capability to analyse the sequence. Plant \& Food Research is part
of the international Potato Genome Sequencing Consortium (PGSC). To
make optimal use of the sequence it needs to be annotated as it is
generated. 

Genome annotation is a complicated, multifaceted procedure. Many
interdependent operations are required to be executed on thousands of
different objects. There are many tools around that can assist in
genome annotation, but most of these tools have one or more drawbacks
concerning installation, capacity and flexibility.

Particularly flexibility is a major problem in bioinformatics as no
two experiments are alike. An common solution is to develop custom
scripts that handle (a part of) a problem. Such scripts should,
ideally be integrated in an analysis pipeline, able to interact with
other, more standardized, building blocks.

Here we describe how to create flexible pipelines, able to embed
customized scripts and capable of processing large amounts of data. We
build these pipelines using a widely available piece of software:
``GNU Make''.

Gnu Make is a commonly used tool to compile software. Compiling
software requires the processing of many interdependent source
files. Such compilation pipelines are described in ``Makefiles'' that
are used by GNU Make to automate the whole process.

We have created a set of generic Makefiles, and some helper scripts,
that describe common operations in genome annotation (amongst others:
Blast, Blat, Glimmer and Bowtie). These Makefiles act as building
blocks of an annotation pipeline. The structure of these Makefiles
allow easy embedding of customized scripts and, hence, the ability to
generate very flexible pipelines.

\section{Usage}

The software aims at providing capable bioinformaticians a framework
that allows them to easily

At the heart of the Makefiles is a library that provides a uniform
interface to using the Makefiles as building blocks in an annotation
pipeline. 


\section{Application}

Using these Makefiles we have build a pipeline (see Figure 1) that has
generated a genome annotation for 1300+ potato BAC sequences.

\section{Conclusion}

\bibliographystyle{plain}
\bibliography{poster}
\end{document}

% LocalWords:  internet Ensembl bioinformaticians Makefile workflow workflows
% LocalWords:  bioinformatics Makefiles Bowtie BAC
