<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
	<title>MOA - Makefiles in bioinformatics</title>
	<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
	<link href="moa.css" rel="stylesheet" type="text/css">
	<link rel="icon" type="image/ico" href="favicon.ico">

  </head>
  <body>
	<center>
	  <div id="wrapper">
		<div style="width: 815px;">
		  <div style="clear: left;">
			<div style="float: left; vertical-align:text-bottom">
			  <img id="menulogo" align="middle" src="images/moa_logo_title.png">
			</div>
			<div style="float: right;">
			  <a href="http://www.plantandfood.co.nz/">
				<img src="images/pfr_logo.jpg" alt="" height="87" width="351" border="0">
			</a>
			</div>
		  </div>
		  <div style="clear: left;">
			<div style="float: left; width: 715px; height: 31px; background-color: rgb(81, 0, 45);">
			  &nbsp;&nbsp;
			</div>
			<div style="float: left;">
			  <img src="images/scoop.jpg" alt="" height="31" width="100" border="0">
			</div>
		  </div>
		</div>
		<div style="clear: left;">
		  <div id="left">
			<div style="width: 160px;">
			  <div style="text-align: right;">
                 <a href="introduction.html">Introduction</a><br>
                 <a href="installation.html">Installation</a><br>
                 <a href="extending.html">Extend Moa</a><br>
                 <a href="reference.html">Reference</a><br>
                 <hr>
                 <a href="moa_manual.pdf">Manual in PDF</a>
                 <hr>
                 <a href="api/index.html">Python API</a><br>
			  </div>
			</div>
		  </div>
		  <div id="center_content">
			<div id="homepage_container">
			
<h1 id="chapterTitle">bowtie</h1>
<p
>Run BOWTIE on an set of input files (query) vs a database index.</p
><div id="targets"
><h1
  >Targets</h1
  ><dl
  ><dt
    >(empty)</dt
    ><dd
    >Execute the default target (bowtie)</dd
    ><dt
    >bowtie</dt
    ><dd
    >(moa_description_bowtie)</dd
    ><dt
    >clean</dt
    ><dd
    >removes all results from this job</dd
    ><dt
    >all</dt
    ><dd
    >executes the default target and into subdirectories to execute any other moa makefile it encounters</dd
    ></dl
  ></div
><div id="parameters"
><h1
  >Parameters</h1
  ><div id="required-parameters"
  ><h2
    >Required parameters</h2
    ><dl
    ><dt
      >title</dt
      ><dd
      >A job name - Describe what you are doing (string)</dd
      ><dt
      >bowtie_db</dt
      ><dd
      >The bowtie database to use. It is allowed to define one of the bowtie database files (.[0-9].ebwt). (file)</dd
      ><dt
      >bowtie_input_dir</dt
      ><dd
      >Input files for bowtie (directory)</dd
      ></dl
    ></div
  ><div id="optional-parameters"
  ><h2
    >Optional parameters</h2
    ><dl
    ><dt
      >moa_precommand</dt
      ><dd
      >A single command to be executed before the main operation starts. For more complicated processing, please override the moa_preprocess target in the local Makefile. (string)</dd
      ><dt
      >moa_postcommand</dt
      ><dd
      >A single shell command to be executed after the Moa is finished. For more complex processing please override the moa_postprocess target in the local Makefile. (string)</dd
      ><dt
      >bowtie_input_extension</dt
      ><dd
      >file extension for the files in bowtie_input_dir (string, default:fastq)</dd
      ><dt
      >bowtie_input_glob</dt
      ><dd
      >glob to select a subset of files from bowtie_input_dir (string, default:*)</dd
      ><dt
      >bowtie_input_sort</dt
      ><dd
      >Sort order. Choose from: u - unsorted, s - size, sr - size reverse, t - time, tr - time reverse ()</dd
      ><dt
      >bowtie_input_limit</dt
      ><dd
      >Number of files to use, if not defined: all files (integer)</dd
      ><dt
      >bowtie_input_format</dt
      ><dd
      >Format of the input files (<strong
	>fastq</strong
	>|fasta)</dd
      ><dt
      >bowtie_extra_params</dt
      ><dd
      >extra parameters to feed bowtie (string)</dd
      ><dt
      >bowtie_paired_ends</dt
      ><dd
      >perform a paired end analysis. If so, the input files are expected to be of the form '*_1.fastq' and ' *_2.fastq' (T|<strong
	>F</strong
	>)</dd
      ><dt
      >bowtie_forward_suffix</dt
      ><dd
      >Last part of the sequence name identifying a file with forward reads (string, default:_1)</dd
      ><dt
      >bowtie_reverse_suffix</dt
      ><dd
      >Last part of the sequence name identifying a file with reverse reads (string, default:_2)</dd
      ><dt
      >bowtie_output_format</dt
      ><dd
      >Format of the output file (bowtie|<strong
	>bam</strong
	>|sam)</dd
      ><dt
      >bowtie_insertsize</dt
      ><dd
      >Expected insertsize (float, default:5000)</dd
      ><dt
      >bowtie_insertsize_sed</dt
      ><dd
      >SED expression to filter the expected insertsize from the input file name (string)</dd
      ><dt
      >bowtie_insertsize_min</dt
      ><dd
      >multiplier determining the minimal acceptable value for two paired reads to be apart. If the bowtie_insertsize is 10000 and this parameter is set at 0.8, than reads that are closer together than 8000 nt are rejecte (float, default:0.1)</dd
      ><dt
      >bowtie_insertsize_max</dt
      ><dd
      >Max insertsize for a paired alignment (float, default:10)</dd
      ></dl
    ></div
  ></div
>
</div>
</div>
  
<p>&nbsp;</p><div class="footer_text" id="footer">&copy 2009 PLANT &amp; FOOD RESEARCH | <a href="http://www.plantandfood.co.nz/disclaimer/">disclaimer</a></div>
</div>
</div></center>


</body>
</html>
